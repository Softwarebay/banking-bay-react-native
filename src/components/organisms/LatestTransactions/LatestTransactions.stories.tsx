import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react-native';
import { ViewStory } from 'utils/storybook/ViewStory';

import {
	LatestTransactions,
	TransactionHistoryItem
} from './LatestTransactions';

const latestTransactions: TransactionHistoryItem[] = [
	{
		id: 'id1',
		iconName: 'smartphone',
		iconColor: '#FF8A71',
		transactionType: 'outgoing',
		title: '+17869871735',
		transactionCategory: 'Mobile Payment',
		amount: 25,
	},
	{
		id: 'id2',
		iconName: 'repeat',
		iconColor: '#55ACEE',
		transactionType: 'incoming',
		title: 'John Snow',
		transactionCategory: 'Money transfer',
		amount: 170.55,
	},
];

storiesOf('Organisms/LatestTransactions', module)
	.addDecorator((getStory) => <ViewStory>{getStory()}</ViewStory>)
	.add('Last two transactions', () => (
		<LatestTransactions
			latestTransactions={latestTransactions}
			onPressHeaderLink={action('onPressHeaderLink')}
		/>
	));
