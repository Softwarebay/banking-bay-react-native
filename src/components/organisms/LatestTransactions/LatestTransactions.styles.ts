import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create({
	card: {
		marginBottom: 6,
	},
	header: {
		justifyContent: 'space-between',
		marginBottom: 14,
	},
});
