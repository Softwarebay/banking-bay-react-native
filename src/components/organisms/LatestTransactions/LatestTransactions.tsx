import { FlatList, StyleProp, View, ViewStyle } from 'react-native';

import { Flex, Heading, IconProps, Link } from 'components/atoms';
import { TransactionCard } from '../TransactionCard/TransactionCard';

import { styles } from './LatestTransactions.styles';

export interface TransactionHistoryItem {
	id: string;
	iconName: IconProps['name'];
	iconColor?: string;
	transactionType: 'incoming' | 'outgoing';
	title: string;
	transactionCategory: string;
	amount: number;
}

interface Props {
	latestTransactions: TransactionHistoryItem[];
	onPressViewAll: () => void;
	onPressTransactionCard: () => void;
	style?: StyleProp<ViewStyle>;
}

export const LatestTransactions = ({
	latestTransactions,
	onPressViewAll,
	onPressTransactionCard,
	style,
}: Props) => {
	return (
		<View style={style}>
			<Flex direction="row" style={styles.header}>
				<Heading
					size="h4"
					testID="latest-transactions-title-test-id"
					text="Latest transactions"
				/>
				<Link
					testID="latest-transaction-link"
					text="View All"
					onPress={onPressViewAll}
				/>
			</Flex>

			<FlatList<TransactionHistoryItem>
				data={latestTransactions}
				renderItem={({
					item: {
						amount,
						iconName,
						id,
						title,
						transactionCategory,
						transactionType,
						iconColor,
					},
					index,
				}) => (
					<TransactionCard
						key={id}
						amount={amount}
						category={transactionCategory}
						iconColor={iconColor}
						iconName={iconName}
						id={id}
						style={[
							styles.card,
							index === 0 && styles.firstCard,
							index === latestTransactions.length - 1 && styles.lastCard,
						]}
						title={title}
						type={transactionType}
						onPress={onPressTransactionCard}
					/>
				)}
			/>
		</View>
	);
};
