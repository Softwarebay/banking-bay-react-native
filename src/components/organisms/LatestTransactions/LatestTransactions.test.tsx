import { fireEvent, render, screen } from '@testing-library/react-native';
import testRenderer from 'react-test-renderer';

import { LatestTransactions } from './LatestTransactions';

import type { Props } from './LatestTransactions';

describe('LatestTransactions', () => {
	const mockOnPressTransactionCard = jest.fn<
		Props['onPressTransactionCard'],
		[]
	>();
	const mockOnPressViewAll = jest.fn<Props['onPressViewAll'], []>();

	const props: Props = {
		onPressTransactionCard: mockOnPressTransactionCard,
		onPressViewAll: mockOnPressViewAll,
		testID: 'latest-transactions-test-id',
		latestTransactions: [
			{
				id: 'item1',
				amount: 2.0,
				iconName: 'amazon',
				title: 'Item1',
				transactionCategory: 'Mobile',
				transactionType: 'outgoing',
			},
			{
				id: 'item2',
				amount: 3.0,
				iconName: 'amazon',
				title: 'Item2',
				transactionCategory: 'Mobile',
				transactionType: 'outgoing',
			},
		],
		style: {
			backgroundColor: 'blue',
		},
	};

	it('should match snapshot', () => {
		const tree = testRenderer
			.create(<LatestTransactions {...props} />)
			.toJSON();

		expect(tree).toMatchSnapshot();
	});

	it("should render section title and 'View all' button", () => {
		render(<LatestTransactions {...props} />);

		const title = screen.getByTestId('latest-transactions-title-test-id');
		const btn = screen.getByTestId('latest-transaction-link-test-id');

		expect(title).toHaveTextContent('Latest transactions');
		expect(btn).toHaveTextContent('View All');
	});

	it("should handle press events on 'View all' button", () => {
		render(<LatestTransactions {...props} />);

		const btn = screen.getByTestId('latest-transaction-link-test-id');

		fireEvent.press(btn);

		expect(mockOnPressViewAll).toHaveBeenCalledTimes(1);
	});

	it('should render latest transactions as a list of transaction cards', () => {
		render(<LatestTransactions {...props} />);

		const transactionItems = screen.getAllByTestId(
			/-latest-transaction-item-test-id/
		);

		expect(transactionItems).toHaveLength(2);
	});

	it('should handle press events on transaction card', () => {
		render(<LatestTransactions {...props} />);

		const transactionItems = screen.getAllByTestId(
			/-latest-transaction-item-test-id/
		);

		transactionItems.forEach((item) => {
			fireEvent.press(item);
		});

		expect(mockOnPressTransactionCard).toHaveBeenCalledTimes(2);
	});

	it('should apply additional styles', () => {
		render(<LatestTransactions {...props} />);

		const latestTransactions = screen.getByTestId(
			'latest-transactions-test-id'
		);

		expect(latestTransactions).toHaveStyle({
			backgroundColor: 'blue',
		});
	});
});
