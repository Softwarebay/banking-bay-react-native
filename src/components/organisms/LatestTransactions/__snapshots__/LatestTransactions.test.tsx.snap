// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LatestTransactions should match snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "blue",
    }
  }
  testID="latest-transactions-test-id"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
        Object {
          "flexDirection": "row",
        },
        Object {
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#1B1D4D",
          },
          Object {
            "fontFamily": "MulishMedium",
            "fontSize": 22,
            "lineHeight": 26.4,
          },
          undefined,
        ]
      }
      testID="latest-transactions-title-test-id"
    >
      Latest transactions
    </Text>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={20}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={false}
      testID="latest-transaction-link-test-id"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FF5887",
              "fontFamily": "MulishMedium",
              "fontSize": 17,
              "lineHeight": 27.2,
            },
            undefined,
          ]
        }
      >
        View All
      </Text>
    </View>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "amount": 2,
          "iconName": "amazon",
          "id": "item1",
          "title": "Item1",
          "transactionCategory": "Mobile",
          "transactionType": "outgoing",
        },
        Object {
          "amount": 3,
          "iconName": "amazon",
          "id": "item2",
          "title": "Item2",
          "transactionCategory": "Mobile",
          "transactionType": "outgoing",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          testID="item1-latest-transaction-item-test-id"
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                },
                Object {
                  "flexDirection": "row",
                },
                Array [
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 10,
                    "shadowColor": "#161966",
                    "shadowOffset": Object {
                      "height": 6,
                      "width": 0,
                    },
                    "shadowOpacity": 0.08,
                    "shadowRadius": 16,
                  },
                  Array [
                    Object {
                      "paddingHorizontal": 15,
                      "paddingVertical": 15,
                    },
                    Array [
                      Object {
                        "marginBottom": 10,
                      },
                      Object {
                        "marginTop": 20,
                      },
                      false,
                    ],
                    false,
                  ],
                ],
              ]
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#F1F5FD",
                  "borderRadius": 100,
                  "height": 50,
                  "justifyContent": "center",
                  "width": 50,
                }
              }
            >
              <SvgMock
                height={25}
                testID="amazon-icon-test-id"
                width={25}
              />
            </View>
            <View
              style={
                Object {
                  "flex": 1,
                  "marginHorizontal": 15,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#1B1D4D",
                    },
                    Object {
                      "fontFamily": "MulishSemiBold",
                      "fontSize": 16,
                      "lineHeight": 25.6,
                    },
                    undefined,
                  ]
                }
                testID="item1-transaction-title-test-id"
              >
                Item1
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "color": "#4C4C60",
                      "fontFamily": "MulishRegular",
                      "fontSize": 17,
                      "lineHeight": 27.2,
                    },
                    Object {
                      "fontSize": 14,
                      "lineHeight": 22.4,
                    },
                  ]
                }
                testID="item1-transaction-subtitle-test-id"
              >
                Mobile
              </Text>
            </View>
            <Text
              style={
                Array [
                  Object {
                    "color": "#1B1D4D",
                  },
                  Object {
                    "fontFamily": "MulishSemiBold",
                    "fontSize": 16,
                    "lineHeight": 25.6,
                  },
                  undefined,
                ]
              }
              testID="item1-transaction-amount-test-id"
            >
              - 2.00
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          testID="item2-latest-transaction-item-test-id"
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                },
                Object {
                  "flexDirection": "row",
                },
                Array [
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 10,
                    "shadowColor": "#161966",
                    "shadowOffset": Object {
                      "height": 6,
                      "width": 0,
                    },
                    "shadowOpacity": 0.08,
                    "shadowRadius": 16,
                  },
                  Array [
                    Object {
                      "paddingHorizontal": 15,
                      "paddingVertical": 15,
                    },
                    Array [
                      Object {
                        "marginBottom": 10,
                      },
                      false,
                      Object {
                        "marginBottom": 20,
                      },
                    ],
                    false,
                  ],
                ],
              ]
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#F1F5FD",
                  "borderRadius": 100,
                  "height": 50,
                  "justifyContent": "center",
                  "width": 50,
                }
              }
            >
              <SvgMock
                height={25}
                testID="amazon-icon-test-id"
                width={25}
              />
            </View>
            <View
              style={
                Object {
                  "flex": 1,
                  "marginHorizontal": 15,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#1B1D4D",
                    },
                    Object {
                      "fontFamily": "MulishSemiBold",
                      "fontSize": 16,
                      "lineHeight": 25.6,
                    },
                    undefined,
                  ]
                }
                testID="item2-transaction-title-test-id"
              >
                Item2
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "color": "#4C4C60",
                      "fontFamily": "MulishRegular",
                      "fontSize": 17,
                      "lineHeight": 27.2,
                    },
                    Object {
                      "fontSize": 14,
                      "lineHeight": 22.4,
                    },
                  ]
                }
                testID="item2-transaction-subtitle-test-id"
              >
                Mobile
              </Text>
            </View>
            <Text
              style={
                Array [
                  Object {
                    "color": "#1B1D4D",
                  },
                  Object {
                    "fontFamily": "MulishSemiBold",
                    "fontSize": 16,
                    "lineHeight": 25.6,
                  },
                  undefined,
                ]
              }
              testID="item2-transaction-amount-test-id"
            >
              - 3.00
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
